Installation issue: curl does not configure/build with Intel compilers spa

@rscohn2

### Steps to reproduce the issue

summary:
`spack install curl@7.80.0 %intel@2021.4.0`

detailed:
```
(base) [myoder96@sh03-05n69 /scratch/users/myoder96/spack_dev/base] (job 39691869) $ spack env create curl2
==> Updating view at /scratch/users/myoder96/spack_dev/base/spack/var/spack/environments/curl2/.spack-env/view
==> Created environment 'curl2' in /scratch/users/myoder96/spack_dev/base/spack/var/spack/environments/curl2
==> You can activate this environment with:
==>   spack env activate curl2
(base) [myoder96@sh03-05n69 /scratch/users/myoder96/spack_dev/base] (job 39691869) $ spack env activate curl2
bash: unalias: despacktivate: not found
(base) [myoder96@sh03-05n69 /scratch/users/myoder96/spack_dev/base] (job 39691869) $ spack add curl@7.80.0 %intel@2021.4.0
(base) [myoder96@sh03-05n69 /scratch/users/myoder96/spack_dev/base] (job 39691869) $ spack concretize --force
==> Starting concretization
==> Environment concretized in 10.35 seconds.
==> Concretized curl@7.80.0%intel@2021.4.0
 -   wmkg47t  curl@7.80.0%intel@2021.4.0 fflags="-fallow-argument-mismatch" ~gssapi~ldap~libidn2~librtmp~libssh~libssh2~nghttp2 patches=bb330614c0d3e046348308fa5dc12132a21a834df2887354fb142516e911d6ba tls=openssl arch=linux-centos7-zen2
[+]  snc5qbt      ^openssl@1.1.1l%intel@2021.4.0 fflags="-fallow-argument-mismatch" ~docs certs=system arch=linux-centos7-zen2
[+]  clrzgmh          ^perl@5.34.0%intel@2021.4.0 fflags="-fallow-argument-mismatch" +cpanm+shared+threads arch=linux-centos7-zen2
[+]  v2znlhh              ^berkeley-db@18.1.40%intel@2021.4.0 fflags="-fallow-argument-mismatch" +cxx~docs+stl patches=b231fcc4d5cff05e5c3a4814f6a5af0e9a966428dc2176540d2c05aff41de522 arch=linux-centos7-zen2
[+]  cr3pw7v              ^bzip2@1.0.8%intel@2021.4.0 fflags="-fallow-argument-mismatch" ~debug~pic+shared arch=linux-centos7-zen2
[+]  656g2tf                  ^diffutils@3.8%intel@2021.4.0 fflags="-fallow-argument-mismatch"  arch=linux-centos7-zen2
[+]  zcaujfu                      ^libiconv@1.16%intel@2021.4.0 fflags="-fallow-argument-mismatch"  libs=shared,static arch=linux-centos7-zen2
[+]  3b6vpfg              ^gdbm@1.19%intel@2021.4.0 fflags="-fallow-argument-mismatch"  arch=linux-centos7-zen2
[+]  miykdke                  ^readline@8.1%intel@2021.4.0 fflags="-fallow-argument-mismatch"  arch=linux-centos7-zen2
[+]  rhs4bbi                      ^ncurses@6.2%intel@2021.4.0 fflags="-fallow-argument-mismatch" ~symlinks+termlib abi=none arch=linux-centos7-zen2
[+]  5bxrapa                          ^pkgconf@1.8.0%intel@2021.4.0 fflags="-fallow-argument-mismatch"  arch=linux-centos7-zen2
[+]  e6kgsmc              ^zlib@1.2.11%intel@2021.4.0 fflags="-fallow-argument-mismatch" +optimize+pic+shared arch=linux-centos7-zen2

==> Updating view at /scratch/users/myoder96/spack_dev/base/spack/var/spack/environments/curl2/.spack-env/view
(base) [myoder96@sh03-05n69 /scratch/users/myoder96/spack_dev/base] (job 39691869) $ spack install
==> Installing environment curl2
[+] /scratch/users/myoder96/spack_dev/base/spack/opt/spack/linux-centos7-zen2/intel-2021.4.0/berkeley-db-18.1.40-v2znlhhcxz6bnphocoyhdlhxpkfb624a
[+] /scratch/users/myoder96/spack_dev/base/spack/opt/spack/linux-centos7-zen2/intel-2021.4.0/libiconv-1.16-zcaujfug2vr3vfeddn7kqxyy275mxijd
[+] /scratch/users/myoder96/spack_dev/base/spack/opt/spack/linux-centos7-zen2/intel-2021.4.0/pkgconf-1.8.0-5bxrapabrixdg3ub6uoeqrq35sz6dqg5
[+] /scratch/users/myoder96/spack_dev/base/spack/opt/spack/linux-centos7-zen2/intel-2021.4.0/zlib-1.2.11-e6kgsmcspij4iwnmz2ncgl34qv4p56wh
[+] /scratch/users/myoder96/spack_dev/base/spack/opt/spack/linux-centos7-zen2/intel-2021.4.0/diffutils-3.8-656g2tfnx2ctb5ev6bkwfrtpke6vrhqp
[+] /scratch/users/myoder96/spack_dev/base/spack/opt/spack/linux-centos7-zen2/intel-2021.4.0/ncurses-6.2-rhs4bbiv63eqpebeoly7igd4z552t4qa
[+] /scratch/users/myoder96/spack_dev/base/spack/opt/spack/linux-centos7-zen2/intel-2021.4.0/bzip2-1.0.8-cr3pw7vvsjueevceokwlnflwa65dccca
[+] /scratch/users/myoder96/spack_dev/base/spack/opt/spack/linux-centos7-zen2/intel-2021.4.0/readline-8.1-miykdkeo4emwqvxr7litxxf2ina7h4hi
[+] /scratch/users/myoder96/spack_dev/base/spack/opt/spack/linux-centos7-zen2/intel-2021.4.0/gdbm-1.19-3b6vpfgwczcfsapaciwnlyfvtpdwb4ao
[+] /scratch/users/myoder96/spack_dev/base/spack/opt/spack/linux-centos7-zen2/intel-2021.4.0/perl-5.34.0-clrzgmhzqbf6n7jehotu5laecfnpko2c
[+] /scratch/users/myoder96/spack_dev/base/spack/opt/spack/linux-centos7-zen2/intel-2021.4.0/openssl-1.1.1l-snc5qbtylltrdy2zl7ahb5vz5ympk5jv
==> Installing curl-7.80.0-wmkg47t77tun2ui5wzrufz72x2x5xfqp
==> No binary for curl-7.80.0-wmkg47t77tun2ui5wzrufz72x2x5xfqp found: installing from source
==> Warning: Intel's compilers may or may not optimize to the same degree for non-Intel microprocessors for optimizations that are not unique to Intel microprocessors
==> Using cached archive: /scratch/users/myoder96/spack_dev/base/spack/var/spack/cache/_source-cache/archive/dd/dd0d150e49cd950aff35e16b628edf04927f0289df42883750cf952bb858189c.tar.bz2
==> Applied patch /scratch/users/myoder96/spack_dev/base/spack/var/spack/repos/builtin/packages/curl/curl_7_78_0_configure_ac.diff
==> curl: Executing phase: 'autoreconf'
==> curl: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/scratch/users/myoder96/spack-build_stage/spack-stage-curl-7.80.0-wmkg47t77tun2ui5wzrufz72x2x5xfqp/spack-src/configure' '--prefix=/scratch/users/myoder96/spack_dev/base/spack/opt/spack/linux-centos7-zen2/intel-2021.4.0/curl-7.80.0-wmkg47t77tun2ui5wzrufz72x2x5xfqp' '--with-zlib=/scratch/users/myoder96/spack_dev/base/spack/opt/spack/linux-centos7-zen2/intel-2021.4.0/zlib-1.2.11-e6kgsmcspij4iwnmz2ncgl34qv4p56wh' '--without-brotli' '--without-libgsasl' '--without-libpsl' '--without-zstd' '--without-ca-bundle' '--without-ca-path' '--with-ca-fallback' '--without-gssapi' '--without-gnutls' '--without-mbedtls' '--without-nss' '--with-openssl=/scratch/users/myoder96/spack_dev/base/spack/opt/spack/linux-centos7-zen2/intel-2021.4.0/openssl-1.1.1l-snc5qbtylltrdy2zl7ahb5vz5ympk5jv' '--without-secure-transport' '--without-libidn2' '--without-librtmp' '--without-nghttp2' '--without-libssh2' '--without-libssh' '--disable-ldap'

1 error found in build log:
     130    configure: compiler options added: -O2
     131    checking if compiler accepts strict warning options... yes
     132    configure: compiler options added: -Wno-system-headers -Wenum-conversion
     133    checking if compiler halts on compilation errors... yes
     134    checking if compiler halts on negative sized arrays... yes
     135    checking if compiler halts on function prototype mismatch... no
  >> 136    configure: error: compiler does not halt on function prototype mismatch.

See build log for details:
  /scratch/users/myoder96/spack-build_stage/spack-stage-curl-7.80.0-wmkg47t77tun2ui5wzrufz72x2x5xfqp/spack-build-out.txt

(base) [myoder96@sh03-05n69 /scratch/users/myoder96/spack_dev/base] (job 39691869) $ 

```

### Information on your system

Stanford's Sherlock HPC:
Linux, CentOS-7

(base) [myoder96@sh03-05n69 /scratch/users/myoder96/spack_dev/base] (job 39691869) $ spack debug report
* **Spack:** 0.17.0-270-c263b64
* **Python:** 3.8.8
* **Platform:** linux-centos7-zen2
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/7622261/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/7622263/spack-build-out.txt)

Comments: The package builds with gcc; this appears to be, fundamentally, a bug with `icc` and a ticket should be opened with Intel (which I should be able to do from my end).

 We can successfully work around the problem by modifying the `configure.ac`, removing CURL_CHECK_COMPILER_PROTOTYPE_MISMATCH  (or writing a patch to do so).
 
 My thought is to patch the package for the short term, and push up a ticket to Intel for the long term?


### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/curl/package.py
+++ b/var/spack/repos/builtin/packages/curl/package.py
@@ -345,6 +345,12 @@ class Curl(NMakePackage, AutotoolsPackage):
     def command(self):
         return Executable(self.prefix.bin.join("curl-config"))
 
+    def flag_handler(self, name, flags):
+        build_system_flags = []
+        if name == "cflags" and self.spec.compiler.name in ["intel", "oneapi"]:
+            build_system_flags = ["-we147"]
+        return flags, None, build_system_flags
+
 
 class AutotoolsBuilder(AutotoolsBuilder):
     def configure_args(self):
@@ -385,10 +391,6 @@ class AutotoolsBuilder(AutotoolsBuilder):
         args += self.with_or_without("libssh2", activation_value="prefix")
         args += self.with_or_without("libssh", activation_value="prefix")
         args += self.enable_or_disable("ldap")
-
-        # https://github.com/spack/spack/issues/27709
-        if self.spec.compiler.name in ["intel", "oneapi"]:
-            args.append("CFLAGS=-we147")
 
         return args
 

--- a/var/spack/repos/builtin/packages/curl/package.py
+++ b/var/spack/repos/builtin/packages/curl/package.py
@@ -386,6 +386,10 @@ class AutotoolsBuilder(AutotoolsBuilder):
         args += self.with_or_without("libssh", activation_value="prefix")
         args += self.enable_or_disable("ldap")
 
+        # https://github.com/spack/spack/issues/27709
+        if self.spec.compiler.name in ["intel", "oneapi"]:
+            args.append("CFLAGS=-we147")
+
         return args
 
     def with_or_without_gnutls(self, activated):

